runner.go

```diff
@@ -148,137 +220,252 @@ func (runner *Runner) Kill() {
 	runner.tomb.Kill(nil)
 }
 
-type workerInfo struct {
-	start        func() (Worker, error)
-	worker       Worker
-	restartDelay time.Duration
-	stopping     bool
+// Worker returns the current worker for the given id.
+// If a worker has been started with the given id but is
+// not currently available, it will wait until it is available,
+// stopping waiting if it receives a value on the stop channel.
+//
+// If there is no worker started with the given id, Worker
+// will return ErrNotFound. If it was stopped while
+// waiting, Worker will return ErrStopped. If the runner
+// has been killed while waiting, Worker will return ErrDying.
+func (runner *Runner) Worker(id string, stop <-chan struct{}) (Worker, error) {
+	runner.mu.Lock()
+	// getWorker returns the current worker for the id
+	// and reports an ErrNotFound error if the worker
+	// isn't found.
+	getWorker := func() (Worker, error) {
+		info := runner.workers[id]
+		if info == nil {
+			// No entry for the id means the worker
+			// will never become available.
+			return nil, ErrNotFound
+		}
+		return info.worker, nil
+	}
+	if w, err := getWorker(); err != nil || w != nil {
+		// The worker is immediately available  (or we know it's
+		// not going to become available). No need
+		// to block waiting for it.
+		runner.mu.Unlock()
+		return w, err
+	}
+	type workerResult struct {
+		w   Worker
+		err error
+	}
+	wc := make(chan workerResult, 1)
+	stopped := false
+	go func() {
+		defer runner.mu.Unlock()
+		for !stopped {
+			runner.workersChangedCond.Wait()
+			if w, err := getWorker(); err != nil || w != nil {
+				wc <- workerResult{w, err}
+				return
+			}
+		}
+	}()
+	select {
+	case w := <-wc:
+		if w.err != nil && errors.Cause(w.err) == ErrNotFound {
+			// If it wasn't found, it's possible that's because
+			// the whole thing has shut down, so
+			// check for dying so that we don't mislead
+			// our caller.
+			select{
+			case <-runner.tomb.Dying():
+				return nil, ErrDying
+			default:
+			}
+		}
+		return w.w, w.err
+	case <-runner.tomb.Dying():
+		return nil, ErrDying
+	case <-stop:
+	}
+	// Stop our wait goroutine.
+	// Strictly speaking this can wake up more waiting Worker calls
+	// than needed, but this shouldn't be a problem as in practice
+	// almost all the time Worker should not need to start the
+	// goroutine.
+	runner.mu.Lock()
+	stopped = true
+	runner.mu.Unlock()
+	runner.workersChangedCond.Broadcast()
+	return nil, ErrStopped
```

Comment:
The issue with the handling of the `mu` and `stopped` in this method is that it's quite hard to reason about when the lock is locked and how/when it's unlocked. I feel like there's got to be better way to structure the code.